File: Chapter1/Chapter1/ViewController.m
//
//  ViewController.m
//  Chapter1
//
//  Created by Jinwoo Kim on 8/26/24.
//

#import "ViewController.h"
#import <MetalKit/MetalKit.h>

@interface ViewController ()
@property (retain, nonatomic) IBOutlet MTKView *mtkView;
@end

@implementation ViewController

- (void)dealloc {
    [_mtkView release];
    [super dealloc];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    id<MTLDevice> device = MTLCreateSystemDefaultDevice();
    assert(device != nil);
    
    self.mtkView.device = device;
    self.mtkView.clearColor = MTLClearColorMake(1., 1., 0.8, 1.);
    self.mtkView.depthStencilPixelFormat = MTLPixelFormatInvalid;
    
    MTKMeshBufferAllocator *allocator = [[MTKMeshBufferAllocator alloc] initWithDevice:device];
    MDLMesh *mdlMesh = [[MDLMesh alloc] initSphereWithExtent:simd_make_float3(0.75, 0.75, 0.75)
                                                    segments:simd_make_uint2(100, 100)
                                               inwardNormals:NO
                                                geometryType:MDLGeometryTypeTriangles
                                                   allocator:allocator];
    [allocator release];
    
    NSError * _Nullable error = nil;
    MTKMesh *mesh = [[MTKMesh alloc] initWithMesh:mdlMesh device:device error:&error];
    [mdlMesh release];
    assert(error == nil);
    
    id<MTLLibrary> library = [device newLibraryWithURL:[NSBundle.mainBundle URLForResource:@"default" withExtension:@"metallib"] error:&error];
    assert(error == nil);
    id<MTLFunction> vertexFunction = [library newFunctionWithName:@"vertex_main"];
    id<MTLFunction> fragmentFunction = [library newFunctionWithName:@"fragment_main"];
    [library release];
    
    MTLRenderPipelineDescriptor *piplineDescriptor = [MTLRenderPipelineDescriptor new];
    piplineDescriptor.colorAttachments[0].pixelFormat = MTLPixelFormatBGRA8Unorm;
    piplineDescriptor.vertexFunction = vertexFunction;
    [vertexFunction release];
    piplineDescriptor.fragmentFunction = fragmentFunction;
    [fragmentFunction release];
    piplineDescriptor.vertexDescriptor = MTKMetalVertexDescriptorFromModelIOWithError(mesh.vertexDescriptor, &error);
    assert(error == nil);
    
    id<MTLRenderPipelineState> pipelineState = [device newRenderPipelineStateWithDescriptor:piplineDescriptor error:&error];
    assert(error == nil);
    [piplineDescriptor release];
    
    //
    
    id<MTLCommandQueue> commandQueue = [device newCommandQueue];
    id<MTLCommandBuffer> commandBuffer = [commandQueue commandBuffer];
    
    MTLRenderPassDescriptor *renderPassDescriptor = [self.mtkView currentRenderPassDescriptor];
    id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:renderPassDescriptor];
    [renderEncoder setRenderPipelineState:pipelineState];
    [pipelineState release];
    
    [renderEncoder setVertexBuffer:mesh.vertexBuffers[0].buffer offset:0 atIndex:0];
    
    [device release];
    
    //
    
    [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                              indexCount:mesh.submeshes[0].indexCount
                               indexType:mesh.submeshes[0].indexType
                             indexBuffer:mesh.submeshes[0].indexBuffer.buffer
                       indexBufferOffset:0];
    [mesh release];
    
    [renderEncoder endEncoding];
    
    //
    
    [commandBuffer presentDrawable:self.mtkView.currentDrawable];
    [commandBuffer commit];
}

@end

File: Chapter1/Chapter1/ViewController.swift
//
//  ViewController.swift
//  Chapter1
//
//  Created by Jinwoo Kim on 7/16/22.
//

import UIKit
import MetalKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Initialize Metal
        guard let device: MTLDevice = MTLCreateSystemDefaultDevice() else {
            fatalError("GPU is not supported")
        }
        
        let frame: CGRect = .init(x: .zero, y: .zero, width: 600, height: 600)
        let mtkView: MTKView = .init(frame: frame, device: device)
        mtkView.clearColor = MTLClearColor(red: 0.4, green: 1, blue: 0.8, alpha: 1)
        view.addSubview(mtkView)
        mtkView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            mtkView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            mtkView.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            mtkView.widthAnchor.constraint(equalToConstant: frame.width),
            mtkView.heightAnchor.constraint(equalToConstant: frame.height)
        ])
        
        // Load a model
        let allocator: MTKMeshBufferAllocator = .init(device: device)
        
        // Sphere
//        let mdlMesh: MDLMesh = .init(sphereWithExtent: [0.75, 0.75, 0.75],
//                                     segments: [100, 100],
//                                     inwardNormals: false,
//                                     geometryType: .triangles,
//                                     allocator: allocator)
        
        // Cone
//        let mdlMesh: MDLMesh = .init(coneWithExtent: [1, 1, 1],
//                                     segments: [10, 10],
//                                     inwardNormals: false,
//                                     cap: true,
//                                     geometryType: .triangles,
//                                     allocator: allocator)
        
        // Train
        let trainUrl: URL = Bundle.main.url(forResource: "train", withExtension: "obj")!

        let vertexDescriptor: MTLVertexDescriptor = .init()
        vertexDescriptor.layouts[0].stride = MemoryLayout<SIMD3<Float>>.stride

        let vertexAttributeDescriptor: MTLVertexAttributeDescriptor = vertexDescriptor.attributes[0]!
        vertexAttributeDescriptor.format = .float3
        vertexAttributeDescriptor.offset = .zero
        vertexAttributeDescriptor.bufferIndex = .zero
        
        let meshDescriptor: MDLVertexDescriptor = try! MTKModelIOVertexDescriptorFromMetalWithError(vertexDescriptor)
        (meshDescriptor.attributes[0] as! MDLVertexAttribute).name = MDLVertexAttributePosition

        let asset: MDLAsset = .init(url: trainUrl, vertexDescriptor: meshDescriptor, bufferAllocator: allocator)
        let mdlMesh: MDLMesh = asset.childObjects(of: MDLMesh.self).first as! MDLMesh
        
        //
        
        let mesh: MTKMesh = try! .init(mesh: mdlMesh, device: device)
        
        // Export to file if needed!
//        let asset: MDLAsset = .init()
//        asset.add(mdlMesh)
//
//        guard MDLAsset.canExportFileExtension("obj") else {
//            fatalError()
//        }
//
//        let baseUrl: URL
//        let objUrl: URL
//        let mtlUrl: URL
//
//        if #available(iOS 16.0, *) {
//            baseUrl = FileManager.default.temporaryDirectory
//            objUrl = baseUrl.appending(path: "primitive").appendingPathExtension("obj")
//            mtlUrl = baseUrl.appending(path: "primitive").appendingPathExtension("mtl")
//        } else {
//            baseUrl = FileManager.default.temporaryDirectory
//            objUrl = baseUrl.appendingPathComponent("primitive").appendingPathExtension("obj")
//            mtlUrl = baseUrl.appendingPathComponent("primitive").appendingPathExtension("mtl")
//        }
//        try! asset.export(to: objUrl)
//
//        print("#####")
//        print(objUrl)
//        print(String(data: try! Data(contentsOf: objUrl), encoding: .utf8)!)
//        print("#####")
//        print(mtlUrl)
//        print(String(data: try! Data(contentsOf: mtlUrl), encoding: .utf8)!)
//        print("#####")
        
        //
        
        guard let commandQueue: MTLCommandQueue = device.makeCommandQueue() else {
            fatalError("Could not create a command queue")
        }
        
        let shaderUrl: URL = Bundle.main.url(forResource: "default", withExtension: "metallib")!
        let library: MTLLibrary = try! device.makeLibrary(URL: shaderUrl)
        let vertexFunction: MTLFunction = library.makeFunction(name: "vertex_main")!
        let fragmentFunction: MTLFunction = library.makeFunction(name: "fragment_main")!
        
        // Set up the pipeline
        
        let pipelineDescriptor: MTLRenderPipelineDescriptor = .init()
        pipelineDescriptor.colorAttachments[0].pixelFormat = .bgra8Unorm
        pipelineDescriptor.vertexFunction = vertexFunction
        pipelineDescriptor.fragmentFunction = fragmentFunction
        
        pipelineDescriptor.vertexDescriptor = try! MTKMetalVertexDescriptorFromModelIOWithError(mesh.vertexDescriptor)
        
        // true
        print(pipelineDescriptor.vertexDescriptor?.isEqual(vertexDescriptor))
        
        let pipelineState: MTLRenderPipelineState = try! device.makeRenderPipelineState(descriptor: pipelineDescriptor)
        
        // Render
        let commandBuffer: MTLCommandBuffer = commandQueue.makeCommandBuffer()!
        let renderPassDescriptor: MTLRenderPassDescriptor = mtkView.currentRenderPassDescriptor!
        let renderEncoder: MTLRenderCommandEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor)!
        
        renderEncoder.setRenderPipelineState(pipelineState)
        renderEncoder.setVertexBuffer(mesh.vertexBuffers[0].buffer, offset: 0, index: 0)
        
//        renderEncoder.setTriangleFillMode(.lines)
        
        mesh.submeshes.forEach { submesh in
            renderEncoder.drawIndexedPrimitives(type: .line,
                                                indexCount: submesh.indexCount,
                                                indexType: submesh.indexType,
                                                indexBuffer: submesh.indexBuffer.buffer,
                                                indexBufferOffset: submesh.indexBuffer.offset)
        }
        
        renderEncoder.endEncoding()
        let drawable: CAMetalDrawable = mtkView.currentDrawable!
        commandBuffer.present(drawable)
        commandBuffer.commit()
    }


}

